lib/AST/Type.cpp

```diff
@@ -902,6 +974,228 @@ Type TypeBase::replaceCovariantResultType(Type newResultType,
+        info |= findExistentialSelfReferences(
+            existentialSig, bgt->getGenericArgs().back(), position);
+      } else {
+        for (auto paramType : bgt->getGenericArgs()) {
+          info |= findExistentialSelfReferences(existentialSig, paramType,
+                                                TypePosition::Invariant);
+        }
+      }
+    }
+
+    return info;
+  }
+
+  // Opaque result types of protocol extension members contain an invariant
+  // reference to 'Self'.
+  if (type->is<OpaqueTypeArchetypeType>())
+    return SelfReferenceInfo::forSelfRef(TypePosition::Invariant);
+
+  // Protocol compositions preserve variance.
+  if (auto *comp = type->getAs<ProtocolCompositionType>()) {
+    // 'Self' may be referenced only in a superclass component.
+    if (const auto superclass = comp->getSuperclass()) {
+      return findExistentialSelfReferences(existentialSig, superclass,
+                                           position);
+    }
+
+    return SelfReferenceInfo();
+  }
+
+  const auto selfTy = existentialSig.getGenericParams().front();
+  if (type->isTypeParameterRootedOn(selfTy)) {
+    const CanType typeInContext =
+        existentialSig.getCanonicalTypeInContext(type);
+    // A direct reference to 'Self'.
+    if (selfTy->isEqual(typeInContext))
+      return SelfReferenceInfo::forSelfRef(position);
+
+    // A reference to an associated type rooted on 'Self'.
+    if (typeInContext->isTypeParameter()) {
+      return SelfReferenceInfo::forAssocTypeRef(position);
+    }
+
+    return findExistentialSelfReferences(existentialSig, typeInContext,
+                                         position);
+  }
+
+  return SelfReferenceInfo();
+}
+
+SelfReferenceInfo TypeBase::findExistentialSelfReferences(
+    const ValueDecl *member,
+    bool treatNonResultCovariantSelfAsInvariant) const {
+  assert(this->isExistentialType());
+
+  // Types never refer to 'Self'.
+  if (isa<TypeDecl>(member))
+    return SelfReferenceInfo();
+
+  auto type = member->getInterfaceType();
+
+  // Skip invalid declarations.
+  if (type->hasError())
+    return SelfReferenceInfo();
+
+  const auto sig = getASTContext().getOpenedArchetypeSignature(
+      Type(const_cast<TypeBase *>(this)));
+
+  if (isa<AbstractFunctionDecl>(member) || isa<SubscriptDecl>(member)) {
+    // For a method, skip the 'self' parameter.
+    if (isa<AbstractFunctionDecl>(member))
+      type = type->castTo<AnyFunctionType>()->getResult();
+
+    auto inputInfo = SelfReferenceInfo();
+    for (auto param : type->castTo<AnyFunctionType>()->getParams()) {
+      // inout parameters are invariant.
+      if (param.isInOut()) {
+        inputInfo |= ::findExistentialSelfReferences(sig, param.getPlainType(),
+                                                     TypePosition::Invariant);
+        continue;
+      }
+      inputInfo |= ::findExistentialSelfReferences(
+          sig, param.getParameterType(), TypePosition::Contravariant);
```

Comment:
This isn't quite right. In `((T) -> U) -> V`, the input `(T) -> U` is contravariant, so `(T)` is covariant and `U` is contravariant in those positions. In general, recurring under the input side of an arrow *flips* variance.

Reply:
This function is the entry point for a member `ValueDecl`. We do flip variance in the recursive `findExistentialSelfReferences` that takes a type and a relative variance position.